Lib/zipimport.py

```diff
@@ -0,0 +1,649 @@
+'''zipimport provides support for importing Python modules from Zip archives.
+
+This module exports three objects:
+- zipimporter: a class; its constructor takes a path to a Zip archive.
+- ZipImportError: exception raised by zipimporter objects. It's a
+  subclass of ImportError, so it can be caught as ImportError, too.
+- _zip_directory_cache: a dict, mapping archive paths to zip directory
+  info dicts, as used in zipimporter._files.
+
+It is usually not needed to use the zipimport module explicitly; it is
+used by the builtin import mechanism for sys.path items that are paths
+to Zip archives.
+'''
+
+#from importlib import _bootstrap_external
+#from importlib import _bootstrap  # for _verbose_message
+import _frozen_importlib_external as _bootstrap_external
+import _frozen_importlib as _bootstrap  # for _verbose_message
+import _imp  # for check_hash_based_pycs
+import _io  # for open
+import marshal  # for loads
+import sys  # for modules
+import time  # for mktime
+
+__all__ = ['ZipImportError', 'zipimporter']
+
+
+path_sep = _bootstrap_external.path_sep
+alt_path_sep = _bootstrap_external.path_separators[1:]
+
+
+class ZipImportError(ImportError):
+    pass
+
+# _read_directory() cache
+_zip_directory_cache = {}
+
+_module_type = type(sys)
+
+
+class zipimporter:
+    '''zipimporter(archivepath) -> zipimporter object
+
+    Create a new zipimporter instance. 'archivepath' must be a path to
+    a zipfile, or to a specific path inside a zipfile. For example, it can be
+    '/tmp/myimport.zip', or '/tmp/myimport.zip/mydirectory', if mydirectory is a
+    valid directory inside the archive.
+
+    'ZipImportError is raised if 'archivepath' doesn't point to a valid Zip
+    archive.
+
+    The 'archive' attribute of zipimporter objects contains the name of the
+    zipfile targeted.
+    '''
+
+    # Split the "subdirectory" from the Zip archive path, lookup a matching
+    # entry in sys.path_importer_cache, fetch the file directory from there
+    # if found, or else read it from the archive.
+    def __init__(self, path):
+        if not isinstance(path, str):
+            import os
+            path = os.fsdecode(path)
+        if not path:
```

Comment:
Nit: Is there a reason to not make this check the first one?

```py
if not path:
    raise ZipImportError('archive path is empty', path=path)
# Check the type of the 'path' later.
```

Reply:
I don't know why this check is needed at first place. I just reproduce the behavior of the C implementation with minimal variance. Supporting path-like objects is new. If there are any problems with an empty string as a path, we should ensure also that the result of `__path__()` is not empty.